src/InstrumentationEngine/ProfilerManager.cpp

```diff
@@ -419,6 +418,192 @@ DWORD WINAPI CProfilerManager::InstrumentationMethodThreadProc(
+                            ::SetEnvironmentVariable(_T("MicrosoftInstrumentationEngine_FileLog"), it->second.c_str());
+                        }
+                        else if (wcscmp(it->first.c_str(), _T("LogFilePath")) == 0)
+                        {
+                            ::SetEnvironmentVariable(_T("MicrosoftInstrumentationEngine_FileLogPath"), it->second.c_str());
+                        }
+                    }
+                }
+            }
+        }
+        else if (wcscmp(bstrCurrChildNodeName, _T("InstrumentationMethods")) == 0)
+        {
+            CComPtr<CXmlNode> pInstrumentationMethodNode;
+            IfFailRet(pCurrChildNode->GetChildNode(&pInstrumentationMethodNode));
+
+            while (pInstrumentationMethodNode != nullptr)
+            {
+                CComBSTR bstrInstrumentationMethodNodeName;
+                IfFailRet(pInstrumentationMethodNode->GetName(&bstrInstrumentationMethodNodeName));
+
+                if (wcscmp(bstrInstrumentationMethodNodeName, _T("AddInstrumentationMethod")) == 0)
+                {
+                    CComBSTR bstrConfigPath;
+                    IfFailRet(pInstrumentationMethodNode->GetAttribute(_T("ConfigPath"), &bstrConfigPath));
+                    IfFalseRet(bstrConfigPath.Length() != 0, E_FAIL);
+
+                    CComPtr<CConfigurationSource> pSource;
+                    pSource.Attach(new (nothrow) CConfigurationSource(bstrConfigPath));
+                    IfFalseRet(nullptr != pSource, E_OUTOFMEMORY);
+
+                    CComPtr<CXmlNode> pInstrumentationMethodChildNode;
+                    IfFailRet(pInstrumentationMethodNode->GetChildNode(&pInstrumentationMethodChildNode));
+
+                    while (pInstrumentationMethodChildNode != nullptr)
+                    {
+                        CComBSTR bstrInstrumentationMethodChildNodeName;
+                        IfFailRet(pInstrumentationMethodChildNode->GetName(&bstrInstrumentationMethodChildNodeName));
+
+                        if (wcscmp(bstrInstrumentationMethodChildNodeName, _T("Settings")) == 0)
+                        {
+                            unordered_map<tstring, tstring> settingsMap;
+                            IfFailRet(ParseSettingsConfigurationNode(pInstrumentationMethodChildNode, settingsMap));
+
+                            for (unordered_map<tstring, tstring>::iterator it = settingsMap.begin();
+                                 it != settingsMap.end();
+                                 it++)
+                            {
+                                IfFailRet(pSource->AddSetting(it->first.c_str(), it->second.c_str()));
+                            }
+                        }
+
+                        CXmlNode* nextInstrumentationMethod = pInstrumentationMethodChildNode->Next();
+                        pInstrumentationMethodChildNode.Release();
+                        pInstrumentationMethodChildNode.Attach(nextInstrumentationMethod);
```

Comment:
>Attach [](start = 56, length = 6)

Attach doesn't increment the ref count. Is that what you want?

Reply:
I believe so. This is the pattern used in ProductionBreakpoints for iterating child nodes.

---
In reply to: [386727155](https://github.com/microsoft/CLRInstrumentationEngine/pull/216#discussion_r386727155) [](ancestors = 386727155)

Reply:
It's weird that you release them after their use, then. It seems that this means the document will point to a bunch of invalid nodes.

---
In reply to: [386727621](https://github.com/microsoft/CLRInstrumentationEngine/pull/216#discussion_r386727621) [](ancestors = 386727621,386727155)

Reply:
Would you mind double-checking the reference counting on these via the debugger? It seems like this could cause mysterious AVs

---
In reply to: [386728314](https://github.com/microsoft/CLRInstrumentationEngine/pull/216#discussion_r386728314) [](ancestors = 386728314,386727621,386727155)

Reply:
That's because Next() news up a new node each time. CXmlNode is an abstraction over the underlying implementation (libxml2 or msxml6) which does the actual navigation to get the next node in the document. This is just throwing away the wrapper once done.

---
In reply to: [386728314](https://github.com/microsoft/CLRInstrumentationEngine/pull/216#discussion_r386728314) [](ancestors = 386728314,386727621,386727155)

Reply:
no problem, I'll double-check

---
In reply to: [386729381](https://github.com/microsoft/CLRInstrumentationEngine/pull/216#discussion_r386729381) [](ancestors = 386729381,386728314,386727621,386727155)

Reply:
I see. That makes sense, then. The normal root for this would be an HRESULT Node::Next(_Out_ Node** pNext), so as to not cause confusion. 

---
In reply to: [386729418](https://github.com/microsoft/CLRInstrumentationEngine/pull/216#discussion_r386729418) [](ancestors = 386729418,386728314,386727621,386727155)